feat: request verifiable presentation from user (#509)

AndrewCLu opened the pull request:
## Explanation

This change implements Verifiable Presentations, a way for users to present Verifiable Credentials to verifiers in an authentic way. The lifecycle is as follows:
1. Web application requests Verifiable Credentials from a user, with a certain request message detailing what type of credentials the service is looking for.
2. A popup allows users to select Verifiable Credentials they wish to share with the application.
3. Cryptkeeper generates a Verifiable Presentation from the selected credentials.
4. Users may either send the Verifiable Presentation without a signature or use their Metamask Wallet to sign the presentation with an Ethereum keypair.
5. The web application verifies that the Verifiable Presentation is authentic and that it contains the credentials it is looking for (not implemented in Cryptkeeper). 

- [x] Create popup allowing users to select Verifiable Credentials
- [x] Generate Verifiable Presentation from selected credentials
- [x] Allow users to sign the Verifiable Presentation
- [x] Connect Verifiable Presentation request and submission to event lifecycle 
- [x] Integrate Verifiable Presentations into demo 
- [x] Add tests for components and hooks
- [x] Make Verifiable Credential Items in popup similar to IdentityList Items

## More Information
See [this change](https://github.com/CryptKeeperZK/crypt-keeper-extension/pull/674) for reference on integration with displaying Verifiable Credentials.

## Screenshots/Screencaps
View a request for Verifiable Credentials and select ones you want to share:
<img width="375" alt="Screen Shot 2023-09-13 at 4 14 33 PM" src="https://github.com/CryptKeeperZK/crypt-keeper-extension/assets/22803219/08cda9d9-056c-41d6-b521-5f7c7354b53f">

Choose to either sign a Verifiable Presentation or send it in the clear:
<img width="381" alt="Screen Shot 2023-09-13 at 4 15 48 PM" src="https://github.com/CryptKeeperZK/crypt-keeper-extension/assets/22803219/9345e803-0c5b-45ae-9ea9-49cf8f15507f">

View a Verifiable Presentation request in full page window:
<img width="1116" alt="Screen Shot 2023-09-13 at 4 16 08 PM" src="https://github.com/CryptKeeperZK/crypt-keeper-extension/assets/22803219/70192f47-52ff-4359-b8d6-119952e20a40">

Sign your Verifiable Presentation with metamask:
<img width="360" alt="Screen Shot 2023-09-11 at 12 11 24 PM" src="https://github.com/CryptKeeperZK/crypt-keeper-extension/assets/22803219/367176ae-e393-43fd-992a-4c47e4830fe5">


## Manual Testing Steps
1. Load the demo
2. Add some Verifiable Credentials to your wallet
3. Click on the button to "Generate a Verifiable Presentation
4. Select the credentials you would like to share
5. Connect your Metamask
6. Authorize the signature
7. Check that the demo has received your signed Verifiable Presentation

## Pre-Merge Checklist

- [x] PR template is filled out
- [x] Pre-commit and pre-push hook checks are passed
- [x] E2E tests are passed locally
- [ ] **IF** this PR fixes a bug, a test that _would have_ caught the bug has been added
- [x] PR is linked to the appropriate GitHub issue
- [x] PR has been added to the appropriate release Milestone

> PR template source from [github.com/MetaMask](https://github.com/MetaMask)

AndrewCLu commented:
@0xisk @0xmad I finished adding tests for the new hooks/components as well as updating the `VerifiableCredentialItem` to match the style of `IdentityItem`, as we discussed in the meeting. It should be ready for a look now.

0xisk commented:
@AndrewCLu Since I have been working on the documentation part, I noticed that we have already two functions published in the provider which are: 
- https://github.com/CryptKeeperZK/crypt-keeper-extension/blob/29b211bae48573f0212ce280fdcc2c4a5932394b/packages/providers/src/sdk/CryptKeeperInjectedProvider.ts#L366-L377 
- https://github.com/CryptKeeperZK/crypt-keeper-extension/blob/29b211bae48573f0212ce280fdcc2c4a5932394b/packages/providers/src/sdk/CryptKeeperInjectedProvider.ts#L379-L389
And now in this PR we have another function: 
- https://github.com/CryptKeeperZK/crypt-keeper-extension/blob/29d4676d5bf299a3283c7045308cfbff9ee3bf45/packages/providers/src/sdk/CryptKeeperInjectedProvider.ts#L393-L407

My worry is to not confuse developers by having non production ready functions in the client. I am thinking of adding a flag inside the provider to distinguish the development features from the others. This is the idea I got but you can thinking of a better way: 

```ts
export class CryptKeeperInjectedProvider {
 const IS_UNDER_DEVELOPMENT = true;

  /**
   * Requests user to add a verifiable credential.
   * NOTE: THIS FUNCTION IS UNDER DEVELOPMENT AND NOT READY FOR PRODUCTION USE
   * 
   * @param {string} serializedVerifiableCredential - The json string representation of the verifiable credential to add.
   * @returns {void}
   */
  async addVerifiableCredentialRequest(serializedVerifiableCredential: string): Promise<void> {
    if (IS_UNDER_DEVELOPMENT) {
      console.warn('This function is under development and not ready for production use.');
      return;
    }
  
    // ...
  }
}
```

We can maybe add _ prefix before the function name. :thinking:

AndrewCLu commented:
> @AndrewCLu Since I have been working on the documentation part, I noticed that we have already two functions published in the provider which are:
> 
> * https://github.com/CryptKeeperZK/crypt-keeper-extension/blob/29b211bae48573f0212ce280fdcc2c4a5932394b/packages/providers/src/sdk/CryptKeeperInjectedProvider.ts#L366-L377
> * https://github.com/CryptKeeperZK/crypt-keeper-extension/blob/29b211bae48573f0212ce280fdcc2c4a5932394b/packages/providers/src/sdk/CryptKeeperInjectedProvider.ts#L379-L389
>   
>   And now in this PR we have another function:
> * https://github.com/CryptKeeperZK/crypt-keeper-extension/blob/29d4676d5bf299a3283c7045308cfbff9ee3bf45/packages/providers/src/sdk/CryptKeeperInjectedProvider.ts#L393-L407
> 
> My worry is to not confuse developers by having non production ready functions in the client. I am thinking of adding a flag inside the provider to distinguish the development features from the others. This is the idea I got but you can thinking of a better way:
> 
> ```ts
> export class CryptKeeperInjectedProvider {
>  const IS_UNDER_DEVELOPMENT = true;
> 
>   /**
>    * Requests user to add a verifiable credential.
>    * NOTE: THIS FUNCTION IS UNDER DEVELOPMENT AND NOT READY FOR PRODUCTION USE
>    * 
>    * @param {string} serializedVerifiableCredential - The json string representation of the verifiable credential to add.
>    * @returns {void}
>    */
>   async addVerifiableCredentialRequest(serializedVerifiableCredential: string): Promise<void> {
>     if (IS_UNDER_DEVELOPMENT) {
>       console.warn('This function is under development and not ready for production use.');
>       return;
>     }
>   
>     // ...
>   }
> }
> ```
> 
> We can maybe add _ prefix before the function name. 🤔

@0xisk @0xmad So right now the `VERIFIABLE_CREDENTIALS` feature flag we use is just for the app package. We could either change it to set up a repo-level config or create a separate flag just for the sdk, but we would have to make sure they are in sync. 

I was thinking that perhaps it isn't necessary to enforce the feature flag at the sdk level, if we do the check in the browser extension anyways. I just renamed the sdk functions with the prefix `DEV_` and added comments. If we think this makes sense, then I will also gate the VC RPC calls with a feature flag check. Let me know what you think.

Also updated the PR with screenshots based on the new UI.

0xmad commented on packages/app/src/background/services/credentials/index.ts:
```diff
@@ -1,4 +1,5 @@
 import { EventName } from "@cryptkeeperzk/providers";
+import { IVerifiablePresentation, IVerifiablePresentationRequest } from "@cryptkeeperzk/types";
```
```suggestion
import type { IVerifiablePresentation, IVerifiablePresentationRequest } from "@cryptkeeperzk/types";
```

0xmad commented on packages/app/src/background/services/credentials/index.ts:
```diff
@@ -199,6 +200,62 @@ export default class VerifiableCredentialsService implements IBackupable {
+    });
+  };
+
+  generateVerifiablePresentation = async (verifiablePresentation: IVerifiablePresentation): Promise<void> => {
+    await this.historyService.trackOperation(OperationType.GENERATE_VERIFIABLE_PRESENTATION, {});
+    await this.notificationService.create({
+      options: {
+        title: "Verifiable Presentation generated.",
+        message: `Generated 1 Verifiable Presentation.`,
+        iconUrl: browser.runtime.getURL("/icons/logo.png"),
+        type: "basic",
+      },
+    });
+    const tabs = await browser.tabs.query({ active: true });
+    await Promise.all(
+      tabs.map((tab) =>
+        browser.tabs
+          .sendMessage(tab.id!, {
+            type: EventName.GENERATE_VERIFIABLE_PRESENTATION,
+            payload: { verifiablePresentation },
+          })
+          .catch(() => undefined),
+      ),
+    );
+  };
```
Why this method has only user notifications? Is it expected that we don't save anything here?

AndrewCLu replied:
Yes. The generated Verifiable Presentation is just emitted as an event to be consumer by the verifier. The extension doesn't need to save anything, except for the history log.

0xmad commented on packages/app/src/background/services/credentials/index.ts:
```diff
@@ -199,6 +200,62 @@ export default class VerifiableCredentialsService implements IBackupable {
     });
   };
 
+  generateVerifiablePresentationRequest = async (
+    verifiablePresentationRequest: IVerifiablePresentationRequest,
+  ): Promise<void> => {
+    await this.browserController.openPopup({
+      params: {
+        redirect: Paths.GENERATE_VERIFIABLE_PRESENTATION_REQUEST,
+        request: verifiablePresentationRequest.request,
+      },
+    });
+  };
```
```suggestion
  generateVerifiablePresentationRequest = async (
    { request }: IVerifiablePresentationRequest,
  ): Promise<void> => {
    await this.browserController.openPopup({
      params: {
        redirect: Paths.GENERATE_VERIFIABLE_PRESENTATION_REQUEST,
        request,
      },
    });
  };
```

0xmad commented on packages/app/src/ui/components/VerifiableCredential/Item/VerifiableCredentialItem.tsx:
```diff
@@ -41,21 +49,50 @@ export const VerifiableCredentialItem = ({
       ? verifiableCredential.issuer
       : verifiableCredential.issuer.id || "unknown";
 
+  const enableSelector = selected !== undefined;
+
+  const enableMenu = onRenameVerifiableCredential !== undefined && onDeleteVerifiableCredential !== undefined;
```
Boolean values shouldn't be verbs.

0xmad commented on packages/app/src/ui/components/VerifiableCredential/Item/VerifiableCredentialItem.tsx:
```diff
@@ -41,21 +49,50 @@ export const VerifiableCredentialItem = ({
+        height: 100,
+        width: "100%",
 
-        "&::-webkit-scrollbar": {
-          display: "none",
```
This style should be added as well.

0xmad commented on packages/app/src/ui/pages/PresentVerifiableCredential/PresentVerifiableCredential.tsx:
```diff
@@ -0,0 +1,51 @@
+import SignVerifiablePresentation from "./components/SignVerifiablePresentation";
+import { usePresentVerifiableCredential } from "./usePresentVerifiableCredential";
+
+const PresentVerifiableCredential = (): JSX.Element => {
```
Component names shouldn't be verbs

0xmad replied:
Sorry, we have this naming convention so let's keep it as current change.

0xmad commented on packages/app/src/ui/pages/PresentVerifiableCredential/PresentVerifiableCredential.tsx:
```diff
@@ -0,0 +1,51 @@
+        onToggleSelectVerifiableCredential={onToggleSelection}
+      />
+    );
+  }
+  return (
```
```suggestion
  }
  
  return (
```

0xmad commented on packages/app/src/ui/pages/PresentVerifiableCredential/__tests__/PresentVerifiableCredential.test.tsx:
```diff
@@ -0,0 +1,144 @@
+ * @jest-environment jsdom
+ */
+
+import { IVerifiablePresentation } from "@cryptkeeperzk/types";
```
```suggestion
import type { IVerifiablePresentation } from "@cryptkeeperzk/types";
```

0xmad commented on packages/app/src/ui/ducks/verifiableCredentials.ts:
```diff
@@ -1,5 +1,6 @@
 /* eslint-disable no-param-reassign */
 import { RPCAction } from "@cryptkeeperzk/providers";
+import { IVerifiablePresentation } from "@cryptkeeperzk/types";
```
```suggestion
import type { IVerifiablePresentation } from "@cryptkeeperzk/types";
```

0xmad commented on packages/app/src/ui/components/VerifiableCredential/Item/useVerifiableCredentialItem.ts:
```diff
@@ -2,7 +2,6 @@ import { useCallback, useState, FormEvent as ReactFormEvent } from "react";
 import { UseFormRegister, useForm } from "react-hook-form";
 
 import { IVerifiableCredentialMetadata } from "@src/types";
```
```suggestion
import type { IVerifiableCredentialMetadata } from "@src/types";
```

0xmad commented on packages/app/src/ui/pages/PresentVerifiableCredential/__tests__/PresentVerifiableCredential.test.tsx:
```diff
@@ -0,0 +1,144 @@
+  };
+
+  beforeEach(() => {
+    createModalRoot();
```
Why do you need modal root here?

AndrewCLu replied:
The two components inside, `VerifiableCredentialSelector` and `VerifiableCredentialSigner`, are rendered within a modal so this is needed for finding the DOM elements

0xmad replied:
It shouldn't use Modal as page container. Take a look at CreateIdentity page as an example.

AndrewCLu replied:
Not sure I understand what you mean. The `CreateIdentity` page test uses a `createModalRoot` as well [here](https://github.com/CryptKeeperZK/crypt-keeper-extension/blob/main/packages/app/src/ui/pages/CreateIdentity/__tests__/CreateIdentity.test.tsx#L81). And the `CreateIdentity` component uses the same Modal components.

0xmad replied:
I will remove this `createModalRoot` for `CreateIdentity` tests in my PR because it doesn't use `FullModal` component inside the page component.

AndrewCLu replied:
Got it. Removed it from this change as well.

0xmad commented on packages/app/src/ui/pages/PresentVerifiableCredential/components/SignVerifiablePresentation/SignVerifiablePresentation.tsx:
```diff
@@ -0,0 +1,97 @@
+              type="submit"
+              variant="contained"
+              onClick={() => {
+                onSubmitVerifiablePresentation(false);
```
It's better to have two separate function than one function with boolean flag.

0xmad commented on packages/app/src/ui/pages/PresentVerifiableCredential/usePresentVerifiableCredential.ts:
```diff
@@ -0,0 +1,176 @@
+  const ethWallet = useEthWallet();
+  const dispatch = useAppDispatch();
+
+  useEffect(() => {
+    function getPresentationRequest() {
+      const { searchParams } = new URL(window.location.href.replace("#", ""));
+      const request = searchParams.get("request");
+
+      if (!request) {
+        return;
+      }
+
+      dispatch(fetchVerifiableCredentials());
+
+      setVerifiablePresentationRequest(request);
+    }
+    getPresentationRequest();
+  }, [setVerifiablePresentationRequest, fetchVerifiableCredentials, dispatch]);
```
This effect won't work properly if url is changed. Also, you don't need to wrap it with function and call it immediately.

AndrewCLu replied:
I fixed it so that there is no function wrap anymore. I'm not sure the Url ever will change in practice. This is the same as for `AddVerifiableCredential` - these are not meant to be screens that are ever shown in their own tab, as we discussed in the last PR, they are simply popups and will be separate from your feature which opens pages in a new tab.

0xmad replied:
If user opens these popups in a new tab manually, it has style issues. In addition, there is a task to support event propagation to multiple tabs which allows user to use the extension in a separate tab or window without closing it. There is no guarantee that it won't be shown in separate tab so I'd fix style issues for this case.

AndrewCLu replied:
Sounds good, addressed in newest changes.

0xmad commented on packages/app/src/ui/pages/PresentVerifiableCredential/usePresentVerifiableCredential.ts:
```diff
@@ -0,0 +1,176 @@
+        return;
+      }
+
+      const address = ethWallet.address?.toLowerCase();
+      const signer = await ethWallet.provider?.getSigner();
+      if (!address || !signer) {
+        setError("Could not connect to Ethereum account.");
+        return;
+      }
```
```suggestion
      const address = ethWallet.address?.toLowerCase();
      const signer = await ethWallet.provider?.getSigner();
   
      if (!address || !signer) {
        setError("Could not connect to Ethereum account.");
        return;
      }
```

0xmad commented on packages/app/src/ui/pages/PresentVerifiableCredential/usePresentVerifiableCredential.ts:
```diff
@@ -0,0 +1,176 @@
+        return;
+      }
+
+      if (!needsSignature) {
```
Why cryptkeeper signature is not supported?

AndrewCLu replied:
As we discussed in the meeting last week, it doesn't make sense for Verifiable Credentials to be signed with a Cryptkeeper signature. These signatures must be verified by a verifier with access to a public registry of keys, and so Ethereum keys makes sense for this purpose. Signing with Cryptkeeper can be implemented if Cryptkeeper keys are included in some public registry

0xmad replied:
Cryptkeeper uses the same standard for wallets and it has the same key pair. If I as a user tries to sign VC with metamask keys, how someone can verify the signature? User doesn't provide public keys in this flow, they just sign the VC.

AndrewCLu replied:
In practice, VCs would be issued to a particular Ethereum address, which would be used as an identifier for the user. Then the signature could just be checked against the address in the VC. Yes, Cryptkeeper keys may be a valid format but I don't expect issuers **and** verifiers to be aware of Cryptkeeper public keys, unless we build a registry.

0xmad replied:
First of all, it needs to support the same user flows. For signing identity user can use both options but for VC only external wallet option. We can understand it internally but how to explain this to users?
Regarding registry: what is the registry for external wallet keys and why do we need to worry about hosting our own registry?

AndrewCLu replied:
Makes sense. Registry is for verifiers to know about public keys. We don't need to make our own, but at least the public keys need to be accessible to a verifier.

0xmad replied:
I created an issue for supporting exporting public keys for cryptkeeper accounts #868

0xmad commented on packages/types/src/index.ts:
```diff
@@ -47,13 +47,14 @@ export type {
 export { ZkProofType } from "./proof";
 export type { IRequestHandler, IPendingRequest } from "./request";
 export { RequestResolutionStatus, PendingRequestType } from "./request";
-export type {
-  IVerifiableCredential,
-  IVerifiablePresentation,
-  ICredentialIssuer,
-  ICredentialSubject,
-  ICredentialStatus,
-  ICredentialProof,
-  ClaimValue,
```
This change should not be reverted

0xmad reviewed (changes requested):
@AndrewCLu thanks, there are some comments.
In addition there is some code which isn't covered with unit tests. It's related to all layers (ui, background, ducks and etc).

0xmad commented:
Regarding ui testing:

1. These styles doesn't match with our identity list styles. It should be aligned to the left side.
![image](https://github.com/CryptKeeperZK/crypt-keeper-extension/assets/14254374/1a7ca992-92e7-447b-9490-d92c2ec6a178)

2. The buttons doesn't match with the previous screen
![image](https://github.com/CryptKeeperZK/crypt-keeper-extension/assets/14254374/2510fb9c-7c9e-42f6-bbe9-4ce9409b5ae1)

3. Why do we need select screen and sign screen if we can use only one screen to complete user action?

4. If I open these pages in a separate tab the styles are broken.
![image](https://github.com/CryptKeeperZK/crypt-keeper-extension/assets/14254374/10640633-0e50-4de2-899e-04223145e454)
![image](https://github.com/CryptKeeperZK/crypt-keeper-extension/assets/14254374/fb28348c-17b4-4c5f-afe1-77782fa6db5a)

AndrewCLu commented:
> Regarding ui testing:

Thanks for the feedback!
1) Fixed the alignment. Updated the screenshots above. 
2) Fixed the button styling differences. Updated the screenshots above.
3) I think the select screen is very important due to the "Reject" button. Being asked to submit credentials is a very intrusive popup, so I think it makes sense to have the user immediately have an option to reject without worrying about selection or signing. If we wanted to include them all on the same page I think it would get too complicated.
4 & 5). Similar to the comment above about URLs, the Verifiable Presentations flow will not ever be experienced by a user in another window. It's just a popup screen requested by the website, and never produced from within the extension. I'm also not quite sure what styling errors you are referring to.

0xmad commented:
@AndrewCLu thanks, there are still some issues:

1. Proceed Without Signing button still doesn't match the height style.
2. If you take a look at `ConnectIdentity` page it works with the same select pattern but here we need to add dropdown button component to handle multiple options for signing. With this approach we can get rid of additional step and keep reject button.
3. If you open two pages you implemented in a separate tab they both have some default page styles, overlay and modal. It's not correct if user works in a separate window. I have a task to support event propagation between the tabs and it allows users to use cryptkeeper between all active tabs where identity is connected. We are not sure if it's always be as-is with popup opening so it's better to fix style errors for this cases. For example, we didn't think we need to create identity from connect screen because there was no connection approach. If we decide to sign VCs using link, we need to support separate screen. It's not a big task and it doesn't require to rewrite the whole logic.

AndrewCLu commented:
> @AndrewCLu thanks, there are still some issues:
> 
> 1. Proceed Without Signing button still doesn't match the height style.
> 2. If you take a look at `ConnectIdentity` page it works with the same select pattern but here we need to add dropdown button component to handle multiple options for signing. With this approach we can get rid of additional step and keep reject button.
> 3. If you open two pages you implemented in a separate tab they both have some default page styles, overlay and modal. It's not correct if user works in a separate window. I have a task to support event propagation between the tabs and it allows users to use cryptkeeper between all active tabs where identity is connected. We are not sure if it's always be as-is with popup opening so it's better to fix style errors for this cases. For example, we didn't think we need to create identity from connect screen because there was no connection approach. If we decide to sign VCs using link, we need to support separate screen. It's not a big task and it doesn't require to rewrite the whole logic.

@0xmad Thanks again for the feedback, I made some changes:
1. Refactored the page so that signing and rejecting happens on the same screen. Now, users can use a dropdown to select which signing method they would like to use, or reject the request. 
2. Created an option to sign with Cryptkeeper. This follows the same UX as `CreateIdentity`.
3. Updated the styles for the popup and separate window. The buttons should now be aligned, and I tested the view on fullscreen. See screenshots for reference. 
4. Removed createModalRoot from tests, as you suggested. 
5. Updated tests to get 100% coverage for the code in these changes (background, ducks, and UI).

0xisk reviewed (changes requested):
Thanks @AndrewCLu, I had my review and left my thoughts and requested changes in this PR https://github.com/CryptKeeperZK/crypt-keeper-extension/pull/867

- Another point: 
Thinking of the PCD idea, why do not we start having a generic interface for defining what are the components of a VC? You can take a look at this package (https://github.com/proofcarryingdata/zupass/blob/main/packages/pcd-types/src/pcd.ts)  to get the idea. More precisely `PCD` and `PCDPackage`

AndrewCLu commented:
@0xmad I updated the styling so that the buttons are now at the bottom. See the new screenshots for reference.

@0xisk I agree with all of your comments and changes in https://github.com/CryptKeeperZK/crypt-keeper-extension/pull/867, but I think its possible for me to separate them into a new PR distinct from this one, since this is centered around verifiable presentations. Perhaps I can combine it with adding comments to all the functions.

0xmad commented on packages/app/src/background/services/credentials/index.ts:
```diff
@@ -201,6 +212,105 @@ export default class VerifiableCredentialsService implements IBackupable {
+        type: "basic",
+      },
+    });
+    const tabs = await browser.tabs.query({ active: true });
+    await Promise.all(
+      tabs.map((tab) =>
+        browser.tabs
+          .sendMessage(tab.id!, {
+            type: EventName.GENERATE_VERIFIABLE_PRESENTATION,
+            payload: { verifiablePresentation: signedVerifiablePresentation },
+          })
+          .catch(() => undefined),
+      ),
+    );
```
Can you reuse here pushEvent method from BrowserUtils? It will propagate the events only for connected host tabs.

AndrewCLu replied:
I think it makes sense to do this with my refactors in the change here: https://github.com/CryptKeeperZK/crypt-keeper-extension/pull/867. I will be changing how all the history/notification/browser events will be made, so I will include this change there.

0xmad commented on packages/app/src/background/services/credentials/index.ts:
```diff
@@ -201,6 +212,105 @@ export default class VerifiableCredentialsService implements IBackupable {
+        type: "basic",
+      },
+    });
+    const tabs = await browser.tabs.query({ active: true });
+    await Promise.all(
+      tabs.map((tab) =>
+        browser.tabs
+          .sendMessage(tab.id!, {
+            type: EventName.USER_REJECT,
+            payload: { type: RejectRequests.VERIFIABLE_PRESENTATION_REQUEST },
+          })
+          .catch(() => undefined),
+      ),
+    );
```
The same is here for pushEvent.

AndrewCLu replied:
Same as above, will include this in my refactors.

0xmad commented on packages/app/src/ui/pages/PresentVerifiableCredential/usePresentVerifiableCredential.ts:
```diff
@@ -0,0 +1,253 @@
+    createVerifiablePresentationFromSelectedCredentials,
+  ]);
+
+  // eslint-disable-next-line consistent-return
```
It's better to return something in the default statement.

0xmad commented on packages/app/src/ui/pages/PresentVerifiableCredential/usePresentVerifiableCredential.ts:
```diff
@@ -0,0 +1,253 @@
+  // eslint-disable-next-line consistent-return
+  const onSubmitVerifiablePresentation = useCallback(async () => {
+    switch (true) {
+      case menuSelectedIndex === 0 && !isWalletConnected:
+        return onConnectWallet();
+      case menuSelectedIndex === 0 && isWalletConnected:
+        return onSubmitVerifiablePresentationWithMetamask();
+      case menuSelectedIndex === 1:
+        return onSubmitVerifiablePresentationWithCryptkeeper();
+      case menuSelectedIndex === 2:
```
These numbers are magic ones. Maybe it's possible to use enum here?

AndrewCLu replied:
fixed

0xmad commented on packages/app/src/ui/pages/PresentVerifiableCredential/PresentVerifiableCredential.tsx:
```diff
@@ -0,0 +1,169 @@
+            Reject
+          </Button>
+
+          <ButtonGroup ref={menuRef} sx={{ height: "2rem", width: "70%" }} variant="contained">
```
Thanks for adding the example how to use it with dropdown. I'll move this one in a separate component and reuse it for create identity.

AndrewCLu replied:
great!

0xmad reviewed (commented):
@AndrewCLu thanks, just some minor comments.

0xmad commented:
@AndrewCLu regarding the test coverage: there are some missing tests for `src/background/services/credentials` folder and for new pages and related components (`AddVerifiableCredential`, `PresentVerifiableCredential`). Usually, I use `test:coverage` report and cover the missing statements, branches according to this report.

0xisk commented:
> @0xmad I updated the styling so that the buttons are now at the bottom. See the new screenshots for reference.
> 
> @0xisk I agree with all of your comments and changes in #867, but I think its possible for me to separate them into a new PR distinct from this one, since this is centered around verifiable presentations. Perhaps I can combine it with adding comments to all the functions.

@AndrewCLu I agree with you I created a PR for this [#876](https://github.com/CryptKeeperZK/crypt-keeper-extension/issues/876)

0xisk reviewed (approved):
Thanks @AndrewCLu I can approve from my end, just please finishing fixing the other requested changes from Anton.

AndrewCLu commented:
> @AndrewCLu regarding the test coverage: there are some missing tests for `src/background/services/credentials` folder and for new pages and related components (`AddVerifiableCredential`, `PresentVerifiableCredential`). Usually, I use `test:coverage` report and cover the missing statements, branches according to this report.

@0xmad Addressed your comments, only caveat is I will include the browser push changes in my [refactor PR](https://github.com/CryptKeeperZK/crypt-keeper-extension/pull/867) since I will be modify all the event related code there anyways.

With regard to test coverage, I just updated all the tests with regards to verifiable presentations (`PresentVerifiableCredential`, those in `services/credential`, etc) to be 100% coverage. I am currently getting my other VC-related code coverage to 100% in the refactor PR which I linked here, since I think it makes sense to keep this change to Verifiable Presentations.

0xmad commented:
@AndrewCLu ok, I'm good with fixing these changes in refactoring PR.
